app/src/test/java/com/kenchen/capstonenewsapp/NewsViewModelTest.kt

```diff
@@ -0,0 +1,145 @@
+        Source(name = "CNN"),
+        "Ken",
+        "Test",
+        "This is a test article",
+        "www.test.com",
+        "www.test.com",
+        "CNN",
+        "Show me how to test.",
+    )
+
+    @ExperimentalCoroutinesApi
+    @get:Rule
+    var coroutinesTestRule = CoroutinesTestRule()
+
+    @get:Rule
+    var instantTaskExecutorRule = InstantTaskExecutorRule()
+
+
+    @Before
+    fun setUp() {
+        MockKAnnotations.init(this)
+
+        // FIXME: can we put every in set up before test start?
+//        every { mockRepository.isDataUsage() } returns flow { emit(true) }
+
+//        viewModel = NewsViewModel(mockRepository)
+    }
+
+    @Test
+    fun `When calling print message actually prints out the message`() {
+
+
+        every { mockRepository.isDataUsage() } returns flow { emit(true) }
+
+        val viewModel = NewsViewModel(mockRepository)
+
+        assertEquals("This is in test", viewModel.getMessage("This is in test"))
+
+        verify(exactly = 1) { mockRepository.isDataUsage() }
+//        verify(exactly = 1) { viewModel.print("123") }
+    }
+
+    @Test
+    fun `print message`() {
+
+        every { mockRepository.isDataUsage() } answers {
+            flow { }
+//            flow { emit(true) }
+        }
+
+        val viewModel = NewsViewModel(mockRepository)
+
+        viewModel.print("1234")
+
+        verify(exactly = 1) { viewModel.print("1234") }
+//        verify(exactly = 1) { println("1234") }
+    }
+
+    @Test
+    fun `fetch articles`() {
+
+        val mockObserver = mockk<Observer<ArticleState>>()
+
+        val expectedResult = ArticleState.Ready(listOf(dummyArticle))
+
+        every { mockRepository.isDataUsage() } returns flow { emit(true) }
+
+        every { mockRepository.getArticles() } returns flow { emit(expectedResult) }
+
+        var onChangedInvoked = 0
+
+
+        val viewModel = NewsViewModel(mockRepository)
+
+        viewModel.headLineNewsLiveData.observeForever(mockObserver)
+
+        every { mockObserver.onChanged(expectedResult) } answers {
+            onChangedInvoked++
+        }
+
+        viewModel.fetchArticle()
+
+        verify(exactly = 1) { mockRepository.getArticles() }
+
+//        assertEquals(1, onChangedInvoked)
+        verify { mockObserver.onChanged(expectedResult) }
+    }
+
+    @Test
+    fun `search articles`() {
+
+        every { mockRepository.isDataUsage() } returns flow { emit(true) }
+
+        val viewModel = NewsViewModel(mockRepository)
+
+        viewModel.searchArticles("test")
+
+        coVerify(exactly = 1) { mockRepository.searchArticles(any()) }
+    }
+
+    @Test
+    fun `toggle data usage`() {
+        every { mockRepository.isDataUsage() } returns flow { emit(true) }
+
+        val viewModel = NewsViewModel(mockRepository)
+
+        viewModel.toggleDataUsage()
+
+        coVerify(exactly = 1) { mockRepository.toggleDataUsage() }
+    }
+}
+
+/**
+ * Question:
+ * 1. how do I test the function that returns Unit (nothing)
```

Comment:
Good questions!

For this one, you would want to ensure and verify that underlying methods are being called.

For example if I have the following:

```kotlin
class A {
    fun aFunction() { ... }
}

class B(val a: A) {
    fun bFunction() {
        a.aFunction()
    }
}
```

Then your test would do this:

```kotlin
val mockA = mockk<A>()
every { mockA.aFunction() } just runs

val sut = B(a)

sut.bFunction()

verify(exactly = 1) { mockA.aFunction() }
```